controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,284 @@
+)
+
+// +kubebuilder:rbac:groups=config.gatekeeper.sh,resources=configs/status,verbs=get;update;list;create;patch
+// +kubebuilder:rbac:groups=constraints.gatekeeper.sh,resources=*,verbs=get;list;watch;create;update;patch;delete
+var (
+	ControllerName = "contraintstatus_reconciler"
+	configGVK      = schema.GroupVersionResource{
+		Group:    v1alpha1.GroupVersion.Group,
+		Version:  v1alpha1.GroupVersion.Version,
+		Resource: "config",
+	}
+)
+
+type discoveryInfo struct {
+	apiResourceList        []*metav1.APIResourceList
+	discoveryLastRefreshed time.Time
+}
+
+type constraintKind struct {
+	apiGroups []string
+	kinds     []string
+}
+
+type ConstraintStatusReconciler struct {
+	client.Client
+	Log                     logr.Logger
+	DynamicClient           dynamic.Interface
+	ClientSet               *kubernetes.Clientset
+	Scheme                  *runtime.Scheme
+	Namespace               string
+	MaxConcurrentReconciles uint
+	discoveryInfo
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ConstraintStatusReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		// The work queue prevents the same item being reconciled concurrently:
+		// https://github.com/kubernetes-sigs/controller-runtime/issues/1416#issuecomment-899833144
+		WithOptions(controller.Options{MaxConcurrentReconciles: int(r.MaxConcurrentReconciles)}).
+		Owns(&v1alpha1.Config{}).
+		Named(ControllerName).
+		For(&v1beta1.ConstraintPodStatus{},
+			builder.WithPredicates(predicate.Funcs{
+				CreateFunc: func(e event.CreateEvent) bool {
+					obj := e.Object.(*v1beta1.ConstraintPodStatus)
+
+					return slices.Contains(obj.Status.Operations, "audit")
+				},
+				UpdateFunc: func(e event.UpdateEvent) bool {
+					oldObj := e.ObjectOld.(*v1beta1.ConstraintPodStatus)
+					newObj := e.ObjectNew.(*v1beta1.ConstraintPodStatus)
+
+					return slices.Contains(newObj.Status.Operations, "audit") &&
+						oldObj.Status.ObservedGeneration != newObj.Status.ObservedGeneration
+				},
+				DeleteFunc: func(e event.DeleteEvent) bool {
+					obj := e.Object.(*v1beta1.ConstraintPodStatus)
+
+					return slices.Contains(obj.Status.Operations, "audit")
+				},
+			},
+			)).
+		Complete(r)
+}
+
+// Reconcile implements reconcile.Reconciler.
+func (r *ConstraintStatusReconciler) Reconcile(ctx context.Context,
+	request reconcile.Request,
+) (reconcile.Result, error) {
+	log := r.Log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	log.Info("Reconciling ConstraintPodStatus and Config")
+
+	gatekeeper := &operatorv1alpha1.Gatekeeper{}
+	err := r.Get(ctx, types.NamespacedName{
+		Namespace: "",
+		Name:      "gatekeeper",
+	}, gatekeeper)
+
+	auditFromCache := *gatekeeper.Spec.Audit.AuditFromCache
+
+	if err != nil || !strings.EqualFold(string(auditFromCache), "Automatic") {
+		return reconcile.Result{Requeue: false}, fmt.Errorf("auditFromCache is not set so skip this reconcile")
+	}
```

Comment:
Would it make sense to not run this controller at all unless `auditFromCache` is set?

Not sure whether this is the best way, but we do this for Gatekeeper syncing, where when Gatekeeper gets installed, the health endpoint fails, causing the pod to restart and revisit the logic on whether the gatekeeper sync controller should run:
- https://github.com/stolostron/governance-policy-framework-addon/blob/cda67fc5101d8cb4b8dd0483ff617c2f37a29b09/controllers/gatekeepersync/health.go#L22-L26
- https://github.com/stolostron/governance-policy-framework-addon/blob/cda67fc5101d8cb4b8dd0483ff617c2f37a29b09/main.go#L607-L610

Reply:
How can we watch auditFromCache?

Reply:
True. I guess the watcher would do exactly this, so there may not really be a benefit to doing that. 🙂

Reply:
I made a comment to this effect in `main.go`. My main concern is when Gatekeeper is not installed, the CRD will be missing for this controller. You could also not have the controller start unless "Automatic" is set as Dale suggested. It would save memory from not caching the ConstraintPodStatus objects